lib/dao.js

```diff
@@ -624,7 +624,184 @@ DataAccessObject.upsert = function(data, options, cb) {
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.patchOrCreateWithWhere =
+  DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+    var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+    if (connectionPromise) { return connectionPromise; }
+    assert(arguments.length >= 1, 'At least one argument is required');
+    if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
+      data = where;
+      where = {};
+    } else if (options === undefined && cb === undefined) {
+      if (typeof data === 'function') {
+        cb = data;
+        data = where;
+        where = {};
+      }
+    } else if (cb === undefined) {
+      if (typeof options === 'function') {
+        cb = options;
+        options = {};
+      }
+    }
+    data = data || {};
+    options = options || {};
+    cb = cb || utils.createPromiseCallback();
+    assert(typeof where === 'object', 'The where argument must be an object');
+    assert(typeof data === 'object', 'The data argument must be an object');
+    assert(typeof options === 'object', 'The options argument must be an object');
+    assert(typeof cb === 'function', 'The cb argument must be a function');
+    var hookState = {};
+    var self = this;
+    var Model = this;
+    var connector = Model.getConnector();
+    var modelObject = connector._models;
+    var modelName = Object.keys(modelObject)[0];
+    var id = getIdValue(this, data);
+    var context = {
+      Model: Model,
+      query: { where: where },
+      hookState: hookState,
+      options: options,
+    };
+    Model.notifyObserversOf('access', context, doUpsertWithNPK);
+    function doUpsertWithNPK(err, ctx) {
+      if (err) return cb(err);
+      var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+      if (!Object.keys(data).length) {
+        var err = new Error('data object cannot be empty!');
+        err.statusCode = 400;
+        process.nextTick(function() { cb(err); });
+        return cb.promise;
+      }
+      if (connector.upsertWithWhere && !isOriginalQuery) {
+        var context = {
+          Model: Model,
+          where: ctx.query.where,
+          data: data,
+          hookState: hookState,
+          options: options,
+        };
+        Model.notifyObserversOf('before save', context, function(err, ctx) {
+          if (err) return cb(err);
+          data = ctx.data;
+          var update = data;
+          var inst = data;
+          if (!(data instanceof Model)) {
+            inst = new Model(data, { applyDefaultValues: false });
+          }
+          update = inst.toObject(false);
+          Model.applyProperties(update, inst);
+          Model = Model.lookupModel(update);
+          var connector = self.getConnector();
```

Comment:
`connector` is already defined on L662, I think this line can be safely removed.